rust/kernel/task.rs

```diff
@@ -196,6 +196,25 @@ impl Task {
         // running.
         unsafe { bindings::wake_up_process(self.0.get()) };
     }
+
+    /// Tries to give up the CPU and lets another task run.
+    ///
+    /// This maps to kernel's `schedule` function, which is similar to
+    /// [`std::thread::yield_now`].
+    ///
+    /// # Context
+    ///
+    /// This function might sleep, don't call in atomic contexts.
+    ///
+    /// [`std::thread::yield_now`]: https://doc.rust-lang.org/std/thread/fn.yield_now.html
+    #[doc(alias = "yield_now")]
+    pub fn schedule() {
+        // SAFETY: If we can schedule back from other thread, then this can be treated as a no-op. A
```

Comment:
Can we violate safety by calling `schedule`?

If we call this from atomic context, we may deadlock the system, which doesn't violate memory safety. If we call it with state set to `TASK_DEAD`, the function never returning doesn't violate safety either, does it? Is there anything else that we could do?

Reply:
It's not about memory objects that the thread owns, it's about the safety of thread users.

We can treat
```
set_special_state(TASK_DEAD);
schedule();
```
, as a decrease on the ref-count of the `task_struct`.

Normally it's OK if we maintain all references to `task_struct` via ref-counts. However, there are a few places in kernel C side where a `task_struct` is referenced with raw pointer. The assumption is that no one should call `do_exit()` or set `TASK_DEAD` unless they are writing `task_struct` or `kthread` code. If Rust code is allowed to  freely `do_exit()` or set `TASK_DEAD`, then C side will have dangling pointers to `task_struct`.

For example, in workqueue, a worker (kernel thread) is referenced without any ref-count addition, so if a `Future` sets `TASK_DEAD` and `schedule`, the worker thread may get freed. And the memory will get corrupted.

Reply:
> It's about the safety of thread users.
> 
> We can treat
> 
> ```
> set_special_state(TASK_DEAD);
> schedule();
> ```
> 
> , as a decrease on the ref-count of the `task_struct`.
> 
> Normally it's OK if we maintain all references to `task_struct` via ref-counts. However, there are a few places in kernel C side where a `task_struct` is referenced with raw pointer. The assumption is that no one should call `do_exit()` or set `TASK_DEAD` unless they are writing `task_struct` or `kthread` code. If Rust code is allowed to freely `do_exit()` or set `TASK_DEAD`, then C side will have dangling pointers to `task_struct`.
> 
> For example, in workqueue, a worker (kernel thread) is referenced without any ref-count addition, so if a `Future` sets `TASK_DEAD` and `schedule`, the worker thread may get freed. And the memory will get corrupted.

I must admit I did think this far originally. It just looked very dangerous to me if we allow Rust code to set the current thread state to `TASK_DEAD`. I will update the `SAFETY:` comments